tools/hexenco.1
.\" Automatically generated by Pod::Man version 1.02
.\" Sat Apr  7 19:06:39 2001
.\"
.\" Standard preamble:
.\" ======================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R

.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used
.\" to do unbreakable dashes and therefore won't be available.  \*(C` and
.\" \*(C' expand to `' in nroff, nothing in troff, for use with C<>
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` `
.    ds C' '
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr
.\" for titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and
.\" index entries marked with X<> in POD.  Of course, you'll have to process
.\" the output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
.    .
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it
.\" makes way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.bd B 3
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ======================================================================
.\"
.IX Title "HEXENCO 1"
.TH HEXENCO 1 "perl v5.6.0" "2001-04-07" "User Contributed Perl Documentation"
.UC
.SH "NAME"
iso2ident, ident2iso: encode/decode a string with alphanumeric chararacters 
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
Given a string \f(CW\*(C`iso2ident\*(C'\fR converts it to a string composed of only
the \f(CW\*(C`[a\-zA\-Z0\-9_]\*(C'\fR set (i.e. alphanumeric + \f(CW\*(C`_\*(C'\fR). Useful for
encoding arbitrary texts in filenames or identifiers. \f(CW\*(C`ident2iso\*(C'\fR
does the inverse task.
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Suppose you want to have files with names \*(L"case: a=0.2345; c=3.56;
b=4.89e-15\*(R". You can do this in Unix but spaces and other characters
may lead to problems. The problem is harder if you want identifiers
for, say, \f(CW\*(C`C\*(C'\fR, \f(CW\*(C`Perl\*(C'\fR or a similar language. This utility converts
all characters different from the set \f(CW\*(C`[a\-zA\-Z0\-9]\*(C'\fR to \f(CW\*(C`_+hexcode\*(C'\fR,
so that you end up with only characters in the set \f(CW\*(C`[a\-zA\-Z0\-9_]\*(C'\fR.
Underscore acts as a 'escape character'.  Of course, the escape
character has to be also encoded. The escape character may be changed
by modifying the value of \f(CW\*(C`$ESCAPE_CHAR\*(C'\fR.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.Vb 5
\& $ iso2ident 'case: a=0.2345; c=3.56; b=4.89e-15'
\& case_3a_20a_3d0_2e2345_3b_20c_3d3_2e56_3b_20b_3d4_2e89e_2d15
\& $ ident2iso case_3a_20a_3d0_2e2345_3b_20c_3d3_2e56_3b_20b_3d4_2e89e_2d15
\& case: a=0.2345; c=3.56; b=4.89e-15
\& $
.Ve
.SH "AUTHOR"
.IX Header "AUTHOR"
Mario A. Storti \f(CW\*(C`mstorti@intec.unl.edu.ar\*(C'\fR
.SH "HISTORYS"
.IX Header "HISTORYS"


ChangeLog
2001-08-07  mstorti@spider  <mstorti@spider>

	* test/lupart/Makefile, test/lupart/check_part.m,
	test/lupart/wallke.epl, test/runtests.pl, src/readmesh.cpp: Adding
	tests for IISD bug

	* test/lupart/Makefile, test/lupart/wallke.epl, src/distmat.cpp,
	src/lusubd.cpp, applications/ns/ns.cpp: Fixed bug for IISD

	* test/lupart/Makefile, src/buffpack.h, src/readmesh.cpp,
	doc/Makefile: Backup prior to incorporate a fix to a bug in IISD.

2001-08-06  mstorti@spider  <mstorti@spider>

	* src/buffpack.h, src/distmat.cpp, src/readmesh.cpp:
	Make workaround for `namespace' not enabled on old gcc
	version in geronimo. (RH 5.2)

	* doc/OBJ/sched2.fig, doc/myexmapl.tex, doc/petscfem.in.tex, doc/pfmat.tex:
	Fixed some bugs in doc

	* tools/coall, test/lupart/Makefile, test/lupart/wallke.epl,
	test/runtests.pl, src/Makefile, src/distmap.h, src/distmat.cpp,
	src/distmat.h, src/elemset.cpp, src/fastmat2.h, src/lusubd.cpp,
	src/pfmat.h, src/readmesh.cpp, applications/ns/wallke.cpp,
	applications/advdif/advective.h, Makefile.base, TODO,
	doc/manual/Makefile, doc/manual/doc.dxx, doc/OBJ/iisd.fig,
	doc/OBJ/iisd2.fig, doc/OBJ/iisd3.fig, doc/OBJ/sched.fig,
	doc/OBJ/sched2.fig, doc/Makefile, doc/idmap.tex, doc/layout.tex,
	doc/odocadvdif.tex, doc/odocelems.tex, doc/odociisd.tex,
	doc/odocns.tex, doc/odocrmsh.tex, doc/petscfem.in.tex,
	doc/pfmat.tex: Added doc for PFMat, IISDMat, DistMap, DistMat and
	DistMatrix.

2001-08-05  mstorti@spider  <mstorti@spider>

	* doc/pfmat.tex:
	Doc for the PFMat abtract interface and in particular de IISD
	implementation.

2001-08-04  mstorti@spider  <mstorti@spider>

	* test/lupart/check_peri.m: Checks the periodic test

	* src/distmat.cpp: *** empty log message ***

2001-08-03  mstorti@spider  <mstorti@spider>

	* VERSION, save.log: Administrative files

	* src/readmesh.cpp.particion:
	Hitch-hiking partitioning was incorporated to `read_mesh'

	* test/lupart/Makefile, test/lupart/mkmesh.m,
	test/lupart/wallke.epl, src/distmap.h, src/distmat.cpp,
	src/distmat.h, src/lusubd.cpp: Fixed a bug in
	DistMap<int,Row>::unpack().  `row' has to be cleared before
	entering values.

	* TODO: order headers

2001-08-02  mstorti@spider  <mstorti@spider>

	* test/distmap/Makefile, test/distmap/distmat.cpp, src/distmap.h,
	src/distmat.cpp, src/distmat.h, src/lusubd.cpp, src/pfmat.h:
	Including the distributed matrix to IISD

	* test/runtests.pl, src/Makefile, src/distmap.h, src/distmat.cpp,
	src/distmat.h, src/lusubd.cpp, src/pfmat.h, src/utils.cpp,
	src/utils.h: Adding distributed matrix class DistMatrix to IISDMat

	* test/distmap/distmat.cpp: Moved code to src/distmat.cpp

	* test/distmap/Makefile: Adding DistMatrix to the IISDMat class.

2001-08-01  mstorti@spider  <mstorti@spider>

	* src/distmat.h: The distributed matrix class

	* src/distmat.cpp: The Distributed Matrix class

	* src/buffpack.h: The BufferPack class allows to pack standard and
	builtins types in a buffer.

	* src/maximizr.h: The maximizer class is a template that remembers
	the maximum of a set of values.

	* src/elemset.cpp, src/lusubd.cpp, src/readmesh.cpp, README,
	README.html: Working on `distmat'.  Backup after incorporating
	changes from some work in geronimo.

	* test/lupart/Makefile, test/distmap/distmat.cpp, src/Makefile, src/readmesh.cpp:
	Working on `distmat'.
	Backup before proceeding to install `distmat' on the IISDMat/PFMat'rices.

	* test/lupart/wallke.epl:
	removed old banner referring to `octave_export_vars'

	* test/distmap/Makefile, test/distmap/distmat.cpp, test/runtests.pl:
	Working on distributed matrix DistMap< int, map<int,double> >

2001-07-31  mstorti@spider  <mstorti@spider>

	* test/distmap/distmat.cpp:
	Distrubuted map<int,map<int,double>> i.e. a sparse matrix

	* test/distmap/distmap.cpp: Fixed a bug

	* tools/myexpect.pl:
	Modified so that without the -o option now sections are not print
	unless at least one test was found.

	* test/distmap/distmap.cpp: Distributed map is working.

	* src/distmap.h:
	Converted `char **buff' for unpacking to `const char *&buff'

	* src/distmap.h:
	Members that do not belong to this process are deleted now.

	* test/Makefile:
	Definition of variables has to be placed before the `includes'
	otherwise the conditional processing ('ifeq' and friend) don't work.

	* src/distmap.h: Distributed map is working basically.

	* src/Makefile: *** empty log message ***

	* test/distmap/Makefile, test/distmap/distmap.cpp, src/distmap.h:
	Debugging the distributed map class.
	Coded the send-receive paired communication.

	* src/distmap.h: Works, paired send-receive.

2001-07-30  mstorti@spider  <mstorti@spider>

	* test/distmap/Makefile, test/distmap/makefile.d, test/Makefile,
	src/distmap.h: Working on distributed map

	* src/Makefile, src/distmap.h, notes.txt, save.log:
	Working on distributed map class.

	* test/distmap/distmap.cpp, test/distmap/makefile.d:
	Workinf on distributed map class. Compiles.

2001-07-28  mstorti@spider  <mstorti@spider>

	* VERSION, save.log: Administrative files

	* test/wkepipe/Makefile, test/wallkechan/Makefile,
	test/wallke/Makefile, test/nonlr/Makefile, test/newff/Makefile,
	test/newff/data.m.epl, test/newff/newff.epl, test/lupart/Makefile,
	test/les/stabi3d/Makefile, test/les/stabi/Makefile,
	test/genload/Makefile, test/bug100/Makefile, test/runtests.pl,
	src/lusubd.cpp, src/pfmat.h, Makefile.base, notes.txt: Found a bug
	in IISD.  Added tests for advdif and ns with IISD.

	* src/lusubd.cpp:
	Fixed bug in lusubd.cpp. Will code it more efficiently.

2001-07-26  mstorti@spider  <mstorti@spider>

	* test/lupart/Makefile, applications/advdif/advdif.cpp, PROJECTS,
	TODO, notes.txt: Porting advdif to PFMat / IISD

2001-07-25  mstorti@spider  <mstorti@spider>

	* src/lusubd.cpp: In geronimo (RH 5.2.) doesn't support
	declarations of the form vector<int> flag0(dofmap->neq,0);

	* test/lupart/Makefile, test/lupart/wallke.epl, applications/ns/nsitetlesfm2.cpp, notes.txt:
	IISD works in multi-processor in RH 7.1 at least.


tools/hexenco.pl
#!/usr/bin/perl

$ESCAPE_CHAR = '_';

sub iso2ident {
    my @line = split "",shift();
    for (@line) {
	$_ = ( /[a-zA-Z0-9]/ ? $_ : $ESCAPE_CHAR.sprintf("%02x",ord($_)));
    }
    return join("",@line);
}

sub ident2iso {
    my @line = split "",shift();
    my @out=();
    while (1) {
	$c = shift @line;
	last unless defined $c;
	$c = chr(hex(shift(@line).shift(@line))) if $c eq $ESCAPE_CHAR;
	push @out,$c;
    }
    return join("",@out);
}

if ($0 =~ /iso2ident/) {
    print iso2ident(shift());
} else {
    print ident2iso(shift());
}

=head1 NAME

iso2ident, ident2iso: encode/decode a string with alphanumeric chararacters 

=head1 SYNOPSIS

Given a string C<iso2ident> converts it to a string composed of only
the C<[a-zA-Z0-9_]> set (i.e. alphanumeric + C<_>). Useful for
encoding arbitrary texts in filenames or identifiers. C<ident2iso>
does the inverse task.

=head1 DESCRIPTION

Suppose you want to have files with names "case: a=0.2345; c=3.56;
b=4.89e-15". You can do this in Unix but spaces and other characters
may lead to problems. The problem is harder if you want identifiers
for, say, C<C>, C<Perl> or a similar language. This utility converts
all characters different from the set C<[a-zA-Z0-9]> to C<_+hexcode>,
so that you end up with only characters in the set C<[a-zA-Z0-9_]>.
Underscore acts as a 'escape character'.  Of course, the escape
character has to be also encoded. The escape character may be changed
by modifying the value of C<$ESCAPE_CHAR>.

=head1 EXAMPLES

 $ iso2ident 'case: a=0.2345; c=3.56; b=4.89e-15'
 case_3a_20a_3d0_2e2345_3b_20c_3d3_2e56_3b_20b_3d4_2e89e_2d15
 $ ident2iso case_3a_20a_3d0_2e2345_3b_20c_3d3_2e56_3b_20b_3d4_2e89e_2d15
 case: a=0.2345; c=3.56; b=4.89e-15
 $ 

=head1 AUTHOR

Mario A. Storti C<mario.storti@gmail.com>

=cut

=cut  
# just for testing
while (<STDIN>) {
    chomp;
    my $t = iso2ident($_);
    print "encoded: \"$t\"\n";
    print "redeencoded: \"",ident2iso($t),"\"\n";
}
=cut
